Library/Homebrew/cmd/mark.rb

```diff
@@ -0,0 +1,101 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "abstract_command"
+require "formula"
+require "tab"
+
+module Homebrew
+  module Cmd
+    class Mark < AbstractCommand
+      cmd_args do
+        description <<~EOS
+          Mark or unmark <formula> as installed on request, or installed as
+          dependency. This can be useful if you want to control whether an
+          installed formula should be removed by `brew autoremove`.
+        EOS
+        switch "--installed-on-request",
+               description: "Mark <formula> as installed on request."
+        switch "--no-installed-on-request",
+               description: "Mark <formula> as not installed on request."
+        switch "--installed-as-dependency",
+               description: "Mark <formula> as installed as dependency."
+        switch "--no-installed-as-dependency",
+               description: "Mark <formula> as not installed as dependency."
+
+        conflicts "--installed-on-request", "--no-installed-on-request"
+        conflicts "--installed-as-dependency", "--no-installed-as-dependency"
```

Comment:
I guess the only issue here is that it doesn't seem to explain _how_ you set a formula to be/to not be auto removed. Of course you and I both know it's about not installed on request and not installed as a dependency but the documentation isn't clear here.

Somewhat similarly: would setting a formula not installed as a dependency mean that `brew autoremove` will attempt to remove it even if it is still an actual dependency?

Perhaps you shouldn't be able to unset `installed_as_dependency` if it is still a dependency (and maybe even not unset `installed_on_request` if it's still a leaf node)?

Reply:
Very interesting points. It makes me wonder if the installed as dependency capabilities are really needed here at all since it could leave the dependency tree in an undesirable state. What is the intended use case of being able to remove or add installed as dependency for a formula?

Reply:
> What is the intended use case of being able to remove or add installed as dependency for a formula?

Same questions from me.

Reply:
> Somewhat similarly: would setting a formula not installed as a dependency mean that `brew autoremove` will attempt to remove it even if it is still an actual dependency?
>
> Perhaps you shouldn't be able to unset `installed_as_dependency` if it is still a dependency (and maybe even not unset `installed_on_request` if it's still a leaf node)?

Having re-read [`utils/autoremove.rb`](https://github.com/Homebrew/brew/blob/7865cc573068790d8e0fb3b4a172e3d2a5e97737/Library/Homebrew/utils/autoremove.rb), I believe `installed_as_dependency` is simply not used when auto-removing (dependency is determined via `Formula#deps`).

So, since changing `installed_on_request` is sufficient to change autoremove behaviour (which, to me, is the main reason for this new command), we can remove the capability of changing `installed_as_dependency`, I think?

Reply:
That makes sense to me.

Reply:
Also makes sense to me. It does, though, make me wonder if this is only really used by `brew autoremove`: would it make sense to be a flag on that? Maybe not, just figured I'd throw that out there.

Reply:
> would it make sense to be a flag on that?

I'd find it a bit confusing if it's implemented that way.